File: ios/TRIX3DCompanion/Core/Services/WardrobeService.swift
//
//  WardrobeService.swift
//  TRIX3DCompanion
//
//  Wardrobe service for avatar/outfit management
//

import Foundation
import Combine

// MARK: - Wardrobe Service Error

enum WardrobeServiceError: Error, LocalizedError {
    case fetchFailed(underlying: Error)
    case equipFailed(underlying: Error)
    case unequipFailed(underlying: Error)
    case notOwned
    case unknown(underlying: Error?)

    var errorDescription: String? {
        switch self {
        case .fetchFailed(let error):
            return "Failed to fetch outfits: \(error.localizedDescription)"
        case .equipFailed(let error):
            return "Failed to equip outfit: \(error.localizedDescription)"
        case .unequipFailed(let error):
            return "Failed to unequip outfit: \(error.localizedDescription)"
        case .notOwned:
            return "You don't own this outfit"
        case .unknown(let error):
            return error?.localizedDescription ?? "Unknown error"
        }
    }
}

// MARK: - Wardrobe Service Protocol

protocol WardrobeServiceProtocol {
    var outfits: [Outfit] { get }
    var isLoading: Bool { get }

    func fetchOutfits() async throws -> [Outfit]
    func equipOutfit(outfitId: String) async throws -> EquipResponse
    func unequipOutfit(outfitId: String) async throws -> EquipResponse
}

// MARK: - Wardrobe Service

@MainActor
final class WardrobeService: ObservableObject, WardrobeServiceProtocol {

    static let shared = WardrobeService()

    // MARK: - Published Properties

    @Published private(set) var outfits: [Outfit] = []
    @Published private(set) var isLoading: Bool = false
    @Published private(set) var lastError: WardrobeServiceError?

    // MARK: - Dependencies

    private let apiClient: APIClient

    // MARK: - Initialization

    init(apiClient: APIClient = .shared) {
        self.apiClient = apiClient
    }

    // MARK: - Public Methods

    /// Fetch all outfits
    func fetchOutfits() async throws -> [Outfit] {
        isLoading = true
        lastError = nil

        do {
            let response: [Outfit] = try await apiClient.get(.wardrobeOutfits)
            self.outfits = response
            isLoading = false
            return response
        } catch {
            let serviceError = WardrobeServiceError.fetchFailed(underlying: error)
            lastError = serviceError
            isLoading = false
            throw serviceError
        }
    }

    /// Equip an outfit
    func equipOutfit(outfitId: String) async throws -> EquipResponse {
        // Check if user owns the outfit
        guard let outfit = outfits.first(where: { $0.id == outfitId }), outfit.isOwned else {
            let error = WardrobeServiceError.notOwned
            lastError = error
            throw error
        }

        isLoading = true
        lastError = nil

        do {
            let response: EquipResponse = try await apiClient.post(.wardrobeEquip(outfitId: outfitId))

            // Update local outfits list
            if response.success {
                self.outfits = self.outfits.map { o in
                    if o.category == outfit.category {
                        return Outfit(
                            id: o.id,
                            name: o.name,
                            category: o.category,
                            image: o.image,
                            previewImage: o.previewImage,
                            isOwned: o.isOwned,
                            isEquipped: o.id == outfitId,
                            description: o.description,
                            price: o.price
                        )
                    }
                    return o
                }
            }

            isLoading = false
            return response
        } catch let error as WardrobeServiceError {
            lastError = error
            isLoading = false
            throw error
        } catch {
            let serviceError = WardrobeServiceError.equipFailed(underlying: error)
            lastError = serviceError
            isLoading = false
            throw serviceError
        }
    }

    /// Unequip an outfit
    func unequipOutfit(outfitId: String) async throws -> EquipResponse {
        isLoading = true
        lastError = nil

        do {
            let response: EquipResponse = try await apiClient.post(.wardrobeUnequip(outfitId: outfitId))

            // Update local outfits list
            if response.success {
                self.outfits = self.outfits.map { o in
                    if o.id == outfitId {
                        return Outfit(
                            id: o.id,
                            name: o.name,
                            category: o.category,
                            image: o.image,
                            previewImage: o.previewImage,
                            isOwned: o.isOwned,
                            isEquipped: false,
                            description: o.description,
                            price: o.price
                        )
                    }
                    return o
                }
            }

            isLoading = false
            return response
        } catch let error as WardrobeServiceError {
            lastError = error
            isLoading = false
            throw error
        } catch {
            let serviceError = WardrobeServiceError.unequipFailed(underlying: error)
            lastError = serviceError
            isLoading = false
            throw serviceError
        }
    }

    // MARK: - Helper Methods

    /// Get outfit by ID
    func outfit(byId id: String) -> Outfit? {
        return outfits.first { $0.id == id }
    }

    /// Get outfits by category
    func outfits(byCategory category: OutfitCategory) -> [Outfit] {
        return outfits.filter { $0.category == category }
    }

    /// Get owned outfits
    func ownedOutfits() -> [Outfit] {
        return outfits.filter { $0.isOwned }
    }

    /// Get equipped outfits
    func equippedOutfits() -> [Outfit] {
        return outfits.filter { $0.isEquipped }
    }

    /// Get outfit for a specific category (the equipped one)
    func equippedOutfit(forCategory category: OutfitCategory) -> Outfit? {
        return outfits.first { $0.category == category && $0.isEquipped }
    }
}

File: src/services/wardrobeService.ts
/**
 * Wardrobe/Avatar system service
 */
import { supabase } from '../config/supabase';
import { logger } from '../utils/logger';
import type {
  Outfit,
  OutfitRow,
  OutfitCategory,
  UserOutfits,
  EquipResponse,
  UnequipResponse,
  OutfitCategoryFilter,
} from '../types/wardrobe';

/**
 * Get all user outfits
 */
export async function getUserOutfits(): Promise<Outfit[]> {
  const { data: { user } } = await supabase.auth.getUser();

  if (!user) {
    throw new Error('请先登录');
  }

  // Get all outfits from database
  const { data: outfits, error } = await supabase
    .from('outfits')
    .select('*')
    .eq('is_active', true)
    .order('category', { ascending: true })
    .order('name', { ascending: true });

  if (error) {
    logger.ui.error('Failed to fetch outfits:', error);
    throw new Error('获取服装列表失败');
  }

  // Get user's owned outfits
  const { data: ownedOutfits } = await supabase
    .from<Array<{ outfit_id: string; is_equipped: boolean }>>('user_outfits')
    .select('outfit_id, is_equipped')
    .eq('user_id', user.id);

  // Get user's equipped outfits
  const equippedIds = new Set(
    (ownedOutfits ?? [])
      .filter((o) => o.is_equipped)
      .map((o) => o.outfit_id)
  );

  const ownedIds = new Set(
    (ownedOutfits ?? []).map((o) => o.outfit_id)
  );

  // Map outfits with ownership status
  const result: Outfit[] = (outfits ?? []).map((outfit: OutfitRow) => ({
    id: outfit.id,
    name: outfit.name,
    category: outfit.category as OutfitCategory,
    image: outfit.image_url,
    previewImage: outfit.preview_image_url || outfit.image_url,
    isOwned: ownedIds.has(outfit.id),
    isEquipped: equippedIds.has(outfit.id),
    description: outfit.description || undefined,
    price: outfit.price || undefined,
  }));

  return result;
}

/**
 * Get outfits filtered by category
 */
export async function getOutfitsByCategory(
  category: OutfitCategoryFilter
): Promise<Outfit[]> {
  const outfits = await getUserOutfits();

  if (category === 'all') {
    return outfits;
  }

  return outfits.filter((outfit) => outfit.category === category);
}

/**
 * Get user's equipped outfits
 */
export async function getEquippedOutfits(): Promise<Outfit[]> {
  const outfits = await getUserOutfits();
  return outfits.filter((outfit) => outfit.isEquipped);
}

/**
 * Equip an outfit
 */
export async function equipOutfit(outfitId: string): Promise<EquipResponse> {
  const { data: { user } } = await supabase.auth.getUser();

  if (!user) {
    return {
      success: false,
      message: '请先登录',
    };
  }

  // Check if user owns this outfit
  const { data: owned } = await supabase
    .from('user_outfits')
    .select('id, is_equipped')
    .eq('user_id', user.id)
    .eq('outfit_id', outfitId)
    .single();

  if (!owned) {
    return {
      success: false,
      message: '您尚未拥有此服装',
    };
  }

  // Get outfit category for uniqueness check
  const { data: outfit } = await supabase
    .from('outfits')
    .select('category')
    .eq('id', outfitId)
    .single();

  if (!outfit) {
    return {
      success: false,
      message: '服装不存在',
    };
  }

  // Unequip other outfits of the same category
  const { data: sameCategoryOutfits } = await supabase
    .from<Array<{ id: string }>>('outfits')
    .select('id')
    .eq('category', outfit.category);

  if (sameCategoryOutfits && sameCategoryOutfits.length > 0) {
    const categoryOutfitIds = sameCategoryOutfits.map((o) => o.id);

    await supabase
      .from('user_outfits')
      .update({ is_equipped: false })
      .eq('user_id', user.id)
      .in('outfit_id', categoryOutfitIds);
  }

  // Equip the selected outfit
  const { error } = await supabase
    .from('user_outfits')
    .update({ is_equipped: true })
    .eq('user_id', user.id)
    .eq('outfit_id', outfitId);

  if (error) {
    logger.ui.error('Failed to equip outfit:', error);
    return {
      success: false,
      message: '装备失败，请重试',
    };
  }

  // Get updated equipped outfits
  const equippedOutfits = await getEquippedOutfits();

  return {
    success: true,
    message: '装备成功',
    equippedOutfits,
  };
}

/**
 * Unequip an outfit
 */
export async function unequipOutfit(outfitId: string): Promise<UnequipResponse> {
  const { data: { user } } = await supabase.auth.getUser();

  if (!user) {
    return {
      success: false,
      message: '请先登录',
    };
  }

  const { error } = await supabase
    .from('user_outfits')
    .update({ is_equipped: false })
    .eq('user_id', user.id)
    .eq('outfit_id', outfitId);

  if (error) {
    logger.ui.error('Failed to unequip outfit:', error);
    return {
      success: false,
      message: '卸下失败，请重试',
    };
  }

  // Get updated equipped outfits
  const equippedOutfits = await getEquippedOutfits();

  return {
    success: true,
    message: '已卸下',
    equippedOutfits,
  };
}

/**
 * Get user's wardrobe summary
 */
export async function getUserWardrobeSummary(): Promise<UserOutfits> {
  const { data: { user } } = await supabase.auth.getUser();

  if (!user) {
    throw new Error('请先登录');
  }

  const outfits = await getUserOutfits();
  const ownedOutfits = outfits.filter((o) => o.isOwned);
  const equippedOutfits = outfits.filter((o) => o.isEquipped);

  return {
    userId: user.id,
    ownedOutfits,
    equippedOutfits,
    totalOutfits: outfits.length,
    ownedCount: ownedOutfits.length,
    equippedCount: equippedOutfits.length,
  };
}
